### PR TITLE
Eo/fix pagination group team

### PR DESCRIPTION
**Pull request by eochoa260**

### CODE REVIEW

File: group_test.go [hunk 1]
```diff
@@ -146,14 +146,98 @@ func TestGetGroupWithAlias(t *testing.T) {
...
+            "description": "third test group",
+            "htmlUrl": "https://app.opslevel-staging.com/groups/test_group_3",
+            "name": "test_group_3",
+            "parent": {
+              "alias": "test_group_1",
+              "id": "Z2lkOi8vb3BzbGV2ZWwvTmFtZXNwYWNlczo6R3JvdXAvMTI"
+            }
+          }
+        ],
+        "pageInfo": {
```
Comment (rocktavious) [inline group_test.go:219]: and this
```suggestion
        {{ template "pagination_second_pageInfo_response" }},
```

File: team.go [hunk 2]
```diff
@@ -343,25 +302,56 @@ func (client *Client) GetTeamCount() (int, error) {
...
+	}
+
+	for q.Account.Teams.PageInfo.HasNextPage {
+		(*variables)["after"] = q.Account.Teams.PageInfo.End
+		resp, err := client.ListTeams(variables)
+		if err != nil {
+			return TeamConnection{}, err
+		}
+		q.Account.Teams.Nodes = append(q.Account.Teams.Nodes, resp.Nodes...)
+		q.Account.Teams.PageInfo = resp.PageInfo
```
Comment (rocktavious) [inline team.go:326]: missing total count summation
```suggestion
		q.Account.Teams.PageInfo = resp.PageInfo
		q.Account.Teams.TotalCount += resp.TotalCount
```

File: team.go [hunk 3]
```diff
@@ -343,25 +302,56 @@ func (client *Client) GetTeamCount() (int, error) {
...
```
Comment (rocktavious) [inline team.go:352]: same total count thing
```suggestion
		q.Account.Teams.PageInfo = resp.PageInfo
		q.Account.Teams.TotalCount += resp.TotalCount
```

File: team.go [hunk 4]
```diff
@@ -343,25 +302,56 @@ func (client *Client) GetTeamCount() (int, error) {
...
 		}
 	}
-	v := client.InitialPageVariables()
-	v["email"] = graphql.String(email)
-	return q.Account.Teams.Query(client, &q, v)
+	if variables == nil {
+		variables = client.InitialPageVariablesPointer()
+	}
+
+	if err := client.Query(&q, *variables); err != nil {
```
Comment (rocktavious) [inline team.go:341]: if you are using the latest `main` branch you can convert this to
```suggestion
	if err := client.Query(&q, *variables, WithName("TeamList")); err != nil {
```
That will then give the query an `operation name` which changes the structure of the query validation but it also means that the Ruby side will beable to emit metrics for it.

File: team_test.go [hunk 5]
```diff
@@ -69,14 +69,250 @@ func TestGetTeamWithAlias(t *testing.T) {
 
 func TestListTeams(t *testing.T) {
 	// Arrange
-	client := ATestClient(t, "team/list")
+	requests := []TestRequest{
+		{`{
+    "query": "query ($after:String!$first:Int!){account{teams(after: $after, first: $first){nodes{alias,id,aliases,contacts{address,displayName,id,type},group{alias,id},htmlUrl,manager{id,email,htmlUrl,name,role},members{nodes{id,email,htmlUrl,name,role},pageInfo{hasNextPage,hasPreviousPage,startCursor,endCursor}},name,responsibilities,tags{nodes{id,key,value},pageInfo{hasNextPage,hasPreviousPage,startCursor,endCursor},totalCount}},pageInfo{hasNextPage,hasPreviousPage,startCursor,endCursor}}}}",
+    "variables": {
```
Comment (rocktavious) [inline team_test.go:75]: same templating adjustments to groups can be applied to teams here.

File: team_test.go [hunk 6]
```diff
@@ -69,14 +69,250 @@ func TestGetTeamWithAlias(t *testing.T) {
...
+          "nodes": [
+            {
+              "alias": "devops",
+              "aliases": [
+                "devops"
+              ],
+              "contacts": [],
+              "htmlUrl": "https://app.opslevel.com/teams/devops",
+              "id": "Z2lkOi8vb3BzbGV2ZWwvVGVhbS83NzQ",
+              "manager": {
```
Comment (rocktavious) [inline team_test.go:93]: As extra credit you could turn some of these things in reusable templates

If you look in `./templates/...` you can see how we define the templates - for example maybe you could make `./templates/user.tpl` and then define

```
{{- define "user_id_1" }}
{
"name": "Kyle Rockman",
"email": "kyle.rockman@mac.com"
}
{{ end }}
```

and then you can convert this to

```suggestion
              "manager": {{ template "user_id_1" }},
```

File: team_test.go [hunk 7]
```diff
@@ -69,14 +69,250 @@ func TestGetTeamWithAlias(t *testing.T) {
...
+                "name": "Kyle Rockman",
+                "email": "kyle.rockman@mac.com"
+              },
+              "members": {
+                "nodes": [
```
Comment (rocktavious) [inline team_test.go:98]: keeping with the template idea you could them pull these out too and convert this to

```suggestion
                "nodes": [
                    {{ template "user_id_2" }},
                    {{ template "user_id_3" }},
                    {{ template "user_id_4" }}
                ]
```

That would help reduce the size and increase the reusability else where.

File: team_test.go [hunk 8]
```diff
@@ -69,14 +69,250 @@ func TestGetTeamWithAlias(t *testing.T) {
...
+  }`},
+	}
+	client := APaginatedTestClient(t, "team/list", requests...)
 	// Act
-	result, err := client.ListTeams()
+	response, err := client.ListTeams(nil)
+	result := response.Nodes
 	// Assert
 	autopilot.Ok(t, err)
 	autopilot.Equals(t, 5, len(result))
```
Comment (rocktavious) [inline team_test.go:311]: user total count here

**Review by rocktavious — CHANGES_REQUESTED** [review]
Great job picking up on the changes and implementing them yourself.  Just a few changes needed a few simplifications that would help.

File: group_test.go [hunk 9]
```diff
@@ -146,14 +146,98 @@ func TestGetGroupWithAlias(t *testing.T) {
 
 func TestListGroups(t *testing.T) {
 	// Arrange
-	client := ATestClient(t, "group/list")
+	requests := []TestRequest{
+		{
+			`{
+    "query": "query ($after:String!$first:Int!){account{groups(after: $after, first: $first){nodes{alias,id,description,htmlUrl,name,parent{alias,id}},pageInfo{hasNextPage,hasPreviousPage,startCursor,endCursor},totalCount}}}",
+    "variables": {
```
[the remark this answers is not in the document]
Comment (eochoa260) [reply group_test.go:153]: Yeah, sorry should have marked this as a draft, ran out of time Friday but was planning on adding that in really quickly today.

File: team_test.go (hunk 8 above)
Comment (eochoa260) [reply team_test.go:311]: Did you mean team total count? I changed it to team total count but I can change it to user total count if that's better.

Comment (rocktavious) [reply team_test.go]: sorry use i mean "use total count here"

**Comment by eochoa260** [comment]
@rocktavious I think this is ready to go

**Review by rocktavious — CHANGES_REQUESTED** [review]
1 missing total count thing it seems

File: team_test.go [hunk 10]
```diff
@@ -69,14 +69,359 @@ func TestGetTeamWithAlias(t *testing.T) {
...
+
+func TestListTeamsWithManager(t *testing.T) {
+	// Arrange
+	requests := []TestRequest{
+		{`{
+    "query": "query TeamList($after:String!$email:String!$first:Int!){account{teams(managerEmail: $email, after: $after, first: $first){nodes{alias,id,aliases,contacts{address,displayName,id,type},group{alias,id},htmlUrl,manager{id,email,htmlUrl,name,role},members{nodes{id,email,htmlUrl,name,role},pageInfo{hasNextPage,hasPreviousPage,startCursor,endCursor},totalCount},name,responsibilities,tags{nodes{id,key,value},pageInfo{hasNextPage,hasPreviousPage,startCursor,endCursor},totalCount}},pageInfo{hasNextPage,hasPreviousPage,startCursor,endCursor},totalCount}}}",
+	"variables": {
+		"after": "",
+		"first": 100,
+		"email": "kyle@opslevel.com"
```
Comment (eochoa260) [inline team_test.go:253]: Had to do this unless we want to change the template for this everywhere

Comment (rocktavious) [reply team_test.go:253]: ahh i see now.  Ya thats fine - kinda the whole point with optionally having templates.  It allows for 1 off's like this

**Review by rocktavious — APPROVED** [review]

**Pull request merged**